### DISCUSSION TITLE
Warning: failure in signing Android apk

### DISCUSSION THREAD

**Discussion by frtykt** [airsdk/Adobe-Runtime-Support#2575] [General]
when i build package apk i getting this error: Warning: failure in signing Android apk
i am using latest harman air sdk
build code: adt -package -target apk -storetype pkcs12 -keystore ihsan.p12 ihsan.apk ihsan-app.xml ihsan.swf

**Comment by OMA2k** [comment]
I also had this problem, and could finally solve it, so I've posted the solution here:
https://github.com/airsdk/Adobe-Runtime-Support/discussions/2659#discussioncomment-5972946

**Comment by frtykt** [comment]
thanks OMA2k i can't do your solusion, i know, i made mistake somewhere, can you help me ? this is so important for me and my customers :(

**Reply by ajwfrost** [reply]
Is your issue the same root cause i.e. is "ihsan.p12" signed with a 1024-bit private key? You may be able to create a new key (2048-bit) and sign the APK with that using the ADT command, and then later on you could re-sign it using "apksigner" (from the Android SDK/tools) using Java 1.7 or early 1.8, plus the "ihsan.p12" file?

thanks

**Reply by OMA2k** [reply]
> thanks OMA2k i can't do your solusion, i know, i made mistake somewhere, can you help me ? this is so important for me and my customers :(

You should describe your problem in detail, then. Otherwise I can't help you.

**Reply by sairope** [reply]
Hey @OMA2k, In june I built and uploaded an aab to the google play store. Now, I'm trying to update my older apps, but I can't even run the apk on my phone cause it's a wrong package due to the 'Warning: Failure in Signing Android apk' error. I don't understand how this could be the case when everything was working fine in June. I'm using Animate CC and latest AIR 52.2.3.4, Java version 8 update 371 (build 1.8.0_371-b11). Should I downgrade to JDK 8u202? I need to sign my app with my old cert and then sign it with a PK in google play to update my old app.

**Comment by OMA2k** [comment]
> Should I downgrade to JDK 8u202?

No, you should use JDK 8u112. Check what I wrote [here](https://github.com/airsdk/Adobe-Runtime-Support/discussions/2659#discussioncomment-5972946).

**Reply by sairope** [reply]
My bad. It was a typo, but I actually don't need to use it since I generated an encrypted certificate zip file using my p12 RSA-1024 cert two eyars ago when the aab was introduced.

So if I understand you correctly what I need to do is:
- upload both encrypted zip of RSA-1024 and RSA-2048 (this will be my upload cert) to the Play Console.
- Now,  I need to use my new RSA-2048 p12 cert to build and sign my app as aab locally, and then I should be able to upload my update to google play without getting any wrong-certificate errors, right?

I will have to upload both RSA-1024 and RSA-2048 encrypted zips every time I want to update my app, correct?

**Reply by marchbold** [reply]
Once you have uploaded your old 1024 cert you won't need to use it again. Google will use it to distribute your app but you should only need to use the upload cert. You should only have to upload the certificates in an encrypted zip once.

**Reply by sairope** [reply]
I have slightly different options to choose from [Screenshot](https://ibb.co/H223HNR):

1. Use existing app signing key from Java keystore
2. Use existing app signing key from another repository
3. Use a new app signing key (this requires ongoing dual releases)

First option seems to be the right option here [Screenshot](https://ibb.co/0q6rH1J).

**Comment by SporeProductions** [comment]
@OMA2k HI. Your instructions over on gitHub were super helpful. I had to read them about 10 times, but I got through all the setup and generated both an encrypted version of my old key and a new pem version of a new 2048 RSA cert. I uploaded the .enc (original) key and Google accepted it, but with this warning:

  _Your app signing key's encryption strength does not meet Google Play's recommended minimum standard_

Your instructions said something about "_upload BOTH"_ the old and the new but there's nowhere to upload the .pem file. My only options on the App Signing page now are:
**Download Certificate** (downloads the app signing key certificate - a **deployment_cert.der** file)
**Request key upgrade** (should I?)
**Download Upload key certificate** (another .der file)
**Request Upload Key reset**

I'm at a loss as to what to do next and what I should be signing my APK with in the publisher settings (using Animate and Air 50.2.3.4)

**Reply by OMA2k** [reply]
Hello. Did you follow the instructions in the step 5 in this screen? https://imgbb.com/H223HNR

If now you can't find a link called "Upload your upload key certificate", you might need to "Request Upload Key reset" to be again able to upload it, since you might have missed that step (which is labeled as "optional" but you have to do it as well in this case). The reset might take 1 or 2 days, IIRC.

Your app should be signed with the "upload_certificate.p12", but before you should upload "upload_certificate.pem" as the "upload key certificate" (different from the old certificate that you also uploaded before) after the upload key has been reset.

I know this is a bit confusing, but I hope you can finally solve it.

**Reply by SporeProductions** [reply]
I did not do step 5, since that step disappeared (I must have skipped it and there's no bringing it back). 

BUT after more digging and trial/error I noticed that the APKs were being rejected for being 32bit. I didn't understand why since I was using AIR 50.x. Turns out Animate 2019 won't show the 64bit build options, and Animate 2020 does, so I moved on to that. 

I built a 32bit APK signed with the OLD p12, then a 64bit APK signed with the OLD p12 too and I was able to upload them and surprisingly submit the release.

**Reply by sairope** [reply]
@SporeProductions How did you generate the .enc file. In my developer console google provides only encryption_public_key.pem now, whereas the old pepk tool uses 64bit encryption key that google used to provide, but it can't read it from the pem file.

**Reply by sairope** [reply]
@OMA2k Unfortunately, google no longer ptovides 'encryption key'. We're getting an 'encryption key path' parameter that requires the .pem file, but this parameter is not supported by the older PEPK tool and results in the following [error](https://ibb.co/jJFCC6M) if you want to use 'encryption key path' with pem file.

I could extract the key from the pem filewith the use of openssl, but does this old pepk tool support '--rsa-aes-encryption' parameter? I tried to use my old 'encryption key' I got from google play back in 2021. Everything packaged without much issue. but then in the console I saw a message stating 'The private key was not properly encrypted, or is not a type of key we support.'

**Reply by SporeProductions** [reply]
> @SporeProductions How did you generate the .enc file. In my developer console google provides only encryption_public_key.pem now, whereas the old pepk tool uses 64bit encryption key that google used to provide, but it can't read it from the pem file.

@sairope 
I used the "specific one" that OMA2k linked to in their instructions. See their comment above where they say _"Check what I wrote here"_ and "**here**" is hyperlinked to the instructions and those instructions in turn contain a hyperlink to the PEPK tool.

Good luck!
